---
title: "Cellular response to Gemcitabine and other drugs"
author: "Diego Montero Solano, Sophie Zhu, Tabita Häßner, Shivani Nundoo"
date: "6/5/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Inhaltsverzeichnis fehlt noch!!

## Loading the data

```{r}
#Set wd to the working directory where this Rscript is located
wd = getwd()

#Load data
basalexpression <- readRDS(paste0(wd,"/CCLE_basalexpression.rds"))
copynumber <- readRDS(paste0(wd,"/CCLE_copynumber.rds"))
mutations <- readRDS(paste0(wd,"/CCLE_mutations.rds"))
treated <- readRDS(paste0(wd,"/NCI_TPW_gep_treated.rds"))
untreated <- readRDS(paste0(wd,"/NCI_TPW_gep_untreated.rds"))
drugsensitivity <- readRDS(paste0(wd,"/NegLogGI50.rds"))

cannotation <- read.table(paste0(wd,"/cellline_annotation.tsv"), header = TRUE, sep ="\t", stringsAsFactors = TRUE)
dannotation <- read.table(paste0(wd,"/drug_annotation.tsv"), header = TRUE, sep ="\t", stringsAsFactors = TRUE)
metadata <- read.table(paste0(wd,"/NCI_TPW_metadata.tsv"), header = TRUE, sep ="\t", stringsAsFactors = TRUE)
```

## Normalization of data

```{r}
#loading library limma needed for normalization
library(limma)

#normalize treated & untreated
untreated_normalized <- normalizeBetweenArrays(untreated)
treated_normalized <- normalizeBetweenArrays(treated)
```

## Comparing boxplots

```{r}
library("viridis")
palette(viridis(15))
drugnames <- metadata$drug
par(mar=c(5,4,5,9))
boxplot(treated,medcol= "black", border=drugnames, col=drugnames, xlab="samples", ylab="gene expression", main="Gene Expression treated celllines", names=FALSE, xaxt="n", boxwex=1, boxlty=0)#
levels=as.factor(levels(drugnames))
legend("topright",inset=c(-0.4,0), legend=levels(drugnames),xpd=TRUE,pch=19,col=levels,title="drugs")

boxplot(treated_normalized, main = "normalized geneexpression after treatment", ylab = "geneexpression", xlab = "samples", names = FALSE) 



```


## PCA

```{r}
#PCA of FC 
fcgeneexpression <- treated_normalized - untreated_normalized
pca_FC = prcomp(fcgeneexpression, center = F, scale. = F)

#plot PC 1-4 of FC

drug_as_factor <- as.factor(metadata[1:ncol(treated),"drug"])
levels <- as.factor(levels(drug_as_factor))
palette(rainbow(15))

par(xpd=T, mar=par()$mar+c(0,0,0,6))
plot(pca_FC$rotation[, 1], pca_FC$rotation[, 2], pch = 20, xlab = "PC 1",ylab = "PC 2", col=drug_as_factor, main = "PC 1&2 of the fold change ")
legend("right",inset = c(-0.3,0), levels(drug_as_factor), xpd = TRUE, pch=20, col = levels)

par(xpd=T, mar=par()$mar+c(0,0,0,5))
plot(pca_FC$rotation[, 3], pca_FC$rotation[, 4], pch = 20, xlab = "PC 3",ylab = "PC 4", col=drug_as_factor, main = "PC 3&4 of the fold change ")
legend("right", inset=c(-0.5,0),levels(drug_as_factor), pch=20, col = levels, xpd =TRUE)
```

FC is obviously drug associated

## T-test

#Normalisation

library("car")


grep("gemcitibine",colnames(treated))

Treated_Cells_Gemcitabine= treated[,365:420]
qqPlot(Treated_Cells_Gemcitabine)



grep("gemcitibine",colnames(untreated))

Untreated_Cells_Gemcitabine=untreated[,365:420]
qqPlot(Treated_Cells_Gemcitabine)



# Checking if normalization will have any effect on the QQ-Plots
library(BBmisc)

Treated_Gemcitabine_Normalised=normalize(Treated_Cells_Gemcitabine)
qqPlot(Treated_Gemcitabine_Normalised)

Untreated_Gemcitabine_Normalised=normalize(Untreated_Cells_Gemcitabine)
qqPlot(Untreated_Gemcitabine_Normalised)

# A more pronounced linearity is observed
# But no significant change in shape, original data is used for further processing.



############T.Test############

# The H0 Hypothesis: there is no change in the mean of the treated and untreated data.
# Alternative Hypothesis (H1): there is a difference in mean between the treated and untreated data.
# Chosen limit: when pValue < 0.05,  H0 is rejected

dim(Untreated_Cells_Gemcitabine)

commondataframe=data.frame(Untreated_Cells_Gemcitabine,Treated_Cells_Gemcitabine)

pValues <- apply(commondataframe, 1, function(x) t.test(x[1:56],x[57:112],paired = TRUE, alternative = "two.sided")$p.value)
commondataframe_pvalue <- data.frame(commondataframe, pValues)
sum(pValues < 0.05)
# 6744 genes had a p value less than 0.05. Thus, this leads us to reject the null hypothesis and assume that there could be a difference between the treated and untreated data for these particular genes.


## Biomarkers

Fold change to find the difference in gene expression between treated and untreated celllines 

```{r}
fcgeneexpression <- treated_normalized - untreated_normalized

#Absolute value of the data in fcgeneexpression, to find the biggest changes, regardless of if they are up- or down-regulated
absvalfcgeneexpression <- abs(fcgeneexpression)
```

New matrices with only the celllines treated with gemcitabine

```{r, warning = FALSE, message = FALSE}
library("dplyr")

treated_gemcitabine <- select(as_tibble(treated_normalized), contains("gemcitibine"))
treated_gemcitabine <- as.matrix(treated_gemcitabine)
rownames(treated_gemcitabine) <- rownames(treated_normalized)

untreated_gemcitabine <- select(as_tibble(untreated_normalized), contains("gemcitibine"))
untreated_gemcitabine <- as.matrix(untreated_gemcitabine)
rownames(untreated_gemcitabine) <- rownames(untreated_normalized)

absvalfcge_gemcitabine <- select(as_tibble(absvalfcgeneexpression), contains("gemcitibine"))
absvalfcge_gemcitabine <- as.matrix(absvalfcge_gemcitabine)
rownames(absvalfcge_gemcitabine) <- rownames(absvalfcgeneexpression)
```


Create a function to order the components of a vector (descending) and return the names in that order

```{r}
orderforvectors <- function(x) {
  y <- names(x[order(-x)])
  return(y)
}
```

Apply the new funtion *orderforvectors* to all the columns and create a new matrix with the gene names

```{r}
matrixofbiomarkers <- apply(absvalfcge_gemcitabine, 2, orderforvectors) 
#the 2 means it is applied to columns
```

Check for most common genes in the first 15 rows of the new matrix

```{r}
topgenes <- sort(table(as.vector(matrixofbiomarkers[1:15,])))
```

Convert *topgenes* into a matrix in descending order of counts

```{r}
matrixtopgenes <- as.matrix(topgenes[order(-topgenes)])
```

Bar plot of *matrixtopgenes* to find the biomarkers

```{r}
barplot(matrixtopgenes[1:20,], xlab = "", ylab = "Counts", main = "Top differentially expressed genes", cex.names = 0.7, las = 2)
mtext("Genes", side = 1, line = 4)
```
We decided to take the top 15 genes as our biomarkers

List of our biomarkers:

```{r}
biomarkers <- names(matrixtopgenes[1:15,1])
biomarkers
```
## Microsatellite instability
#MSI is defined as a mistake in MMR (Mismatch Repair system) so if we plot the number of mutations against the celllines, MSI-H celllines are supposed to have more mutations, to check this, we do the following plot

#Plotting mutations and Counting mutations -> MSI relation?

```{r echo=TRUE, results = 'hide', warning=FALSE, message = F}
library(plyr)
library(qpcR)
library(ggplot2)
```
```{r}
mutation_cellline <- mutations[,16]
counting_mutations <- count(mutation_cellline, vars = NULL, wt_var = NULL)
```

```{r}
#creating a new dataframe with the Celllines and whether the Microsatellite Instablility is high or not. TRUE:High  False:Stable/Low

L=cannotation$Microsatellite_instability_status=="MSI-H"
Cell_Line=data.frame("Cell Line name"= cannotation[1:61,1],"Microsatellite"= L)
Cell_Line <- Cell_Line[order(Cell_Line$Cell.Line.name),]
drugsensitivity=t(drugsensitivity)

# It is observed that the cell lines in the drugsensitivity matrix is already arranged alphabetically.
# Thus the Cell_Line dataframe is directly merged with the Gemcitabine Drug Sensitivity Values.
Cell_Line_no_nas <- na.omit(Cell_Line)
table_of_mutations <- qpcR:::cbind.na(counting_mutations, Cell_Line_no_nas)

mutation_plot <- qplot(

    x = table_of_mutations[,3],

    y = table_of_mutations[,2],

    data = table_of_mutations,

    color = Microsatellite, 

    main = "number of mutations per cellline coloured by MSI status", 

    xlab = "Cell line",ylab = "mutations",
)
mutation_plot + ggtitle("number of mutations per cellline coloures by MSI status")+ xlab("Cell Line")+ylab("mutations")+theme(axis.text.x = element_text(angle = 90))

```
It is obvious that MSI status affects the number of mutations, due to mutations in the repair system, but does MSI status also effect the effect of gemcitibine? MSI-H cancers are known to have a better prognosis when it comes to immune activating therapies, because of more unknown antigens, being produced. But does it also effect non-immune mechanisms in the cell? A loss of system which repairs wrong parts could also lead to a worse proliferation of the cell.

```{r, warning=FALSE}
# Plotting the GI50 values relating to Gemcitabine treatment .


#creating a new dataframe with the Celllines and whether the Microsatellite Instablility is high or not. TRUE:High  False:Stable/Low

L=cannotation$Microsatellite_instability_status=="MSI-H"
Cell_Line=data.frame("Cell Line name"= cannotation[1:61,1],"Microsatellite"= L)
Cell_Line <- Cell_Line[order(Cell_Line$Cell.Line.name),]
drugsensitivity=t(drugsensitivity)

# It is observed that the cell lines in the drugsensitivity matrix is already arranged alphabetically.
# Thus the Cell_Line dataframe is directly merged with the Gemcitabine Drug Sensitivity Values.

gemcitabine_line=data.frame("Gemcitabine drug sensitivity"= drugsensitivity[8,1:61])
everything=cbind(Cell_Line,gemcitabine_line)

library(ggplot2)
basic=ggplot(everything, aes(Cell.Line.name, Gemcitabine.drug.sensitivity, colour = factor(Microsatellite), shape = factor(Microsatellite) )) +
   geom_point()
basic + ggtitle("Drug Sensitivity of Gemcitabine treated cell lines according to High Microsatellite Instability")+ xlab("Cell Line")+ylab("Drug Sensitivity")+theme(axis.text.x = element_text(angle = 90))

# a relationship between high drug sensitivity(Gemcitabine) and high microsatellite instability cannot be concluded.
```

#Heatmap of biomarkers and of MSI related genes 
To answer the question wheter MSI status affects the fold change after gemcitibine treatment, we investigate the fold change of our 15 biomarkers for gemcitibine.

```{r}
#check which rows are for gemcitibine
gmci <- grepl('gemcitibine', colnames(fcgeneexpression))

#rows 365:420(gemcitibine), make dataframe with gemci and biomarkers
colnames(fcgeneexpression) = metadata[1:819,2]
biomarkers_geneexpression <- fcgeneexpression[c("ATF3", "CXCL8", "GADD45A", "CDKN1A", "TNFAIP3", "EGR1", "GDF15", "BTG2", "GEM", "CEACAM1", "DHRS2", "HBEGF", "IL11", "TNFRSF9", "TXNIP"),365:420]

#MSI related genes, choosen after research, new matrix for untreated and FC
geneexpression_MSI_related <- fcgeneexpression[c("MLH3", "MLH1", "MSH6", "PMS2CL///PMS2", "BATF", "BATF3"), 365:420]
geneexpression_MSI_related_untreated <- untreated_normalized[c("MLH3", "MLH1", "MSH6", "BATF", "BATF3"), 365:420]


#which cellines MSI-H or L? Note them 
#View(Cell_Line)

#MSI-H in col = 8,12,11,17,19,27,42,45 of geneexpression MSI related which is only gemcitibine specific


#define new matrices to divide matrices in MSI-H and MSI-L

biomarkers_geneexpression_MSIH <- biomarkers_geneexpression[,c(8,12,11,17,19,27,42,45)]
biomarkers_geneexpression_MSIL <- biomarkers_geneexpression[,-c(8,12,11,17,19,27,42,45)]

geneexpression_MSI_related_MSIH <- geneexpression_MSI_related[,c(8,12,11,17,19,27,42,45)]
geneexpression_MSI_related_MSIL <- geneexpression_MSI_related[,-c(8,12,11,17,19,27,42,45)]

geneexpression_MSI_related_untreated_MSIH <- geneexpression_MSI_related_untreated[, c(8,12,11,17,19,27,42,45)]
geneexpression_MSI_related_untreated_MSIL <- geneexpression_MSI_related_untreated[, -c(8,12,11,17,19,27,42,45)]


#packages which makes it possible to plot tow heatmaps next to each other
```
```{r echo=TRUE, results = 'hide', warning=FALSE, message = F}
library("ComplexHeatmap")
library("circlize")
library("colorspace")
library("GetoptLong")
```
Does MSI status effect the expression of biomarkers after gemcitibine treatment?
```{r}
#two heatmaps of FC of biomarkers in MSI-Low and MSI-High celllines
ht_biomarkers_geneexpression_MSIH = Heatmap(biomarkers_geneexpression_MSIH, name = "FC MSI-High", column_title = "MSI-H", column_names_gp = gpar(fontsize = 7)) 
ht_biomarkers_geneexpression_MSIL = Heatmap(biomarkers_geneexpression_MSIL, name = "FC MSI-Low", column_title = "MSI-L", column_names_gp = gpar(fontsize = 7))
 
ht_biomarkers_geneexpression_MSIH + ht_biomarkers_geneexpression_MSIL


#no obvious difference between MSI-H and MSI-L when it comes to FC after gemci treatment
#sum mean of FC in MSI-L (48 celllines) and MSI-H (8 celllines)

mean(biomarkers_geneexpression_MSIH) ;mean(biomarkers_geneexpression_MSIL)

```
no significant difference between geneexpression in MSI-H and MSI-L biomarkers

##Drug sensitivity and cell line

```{r}
#is the drug sensitivity for gemcitabine dependend on the cell line? 
#plotting mean graph plot 

#calculation of GI50 mean for gemcitabine
rowMeans(drugsensitivity, na.rm = TRUE, dims = 1)
#GI50 mean value is 7,003322 for gemcitabine
#creating a vector for GI50 values of gemcitabine only 
NegLogGI50gemcitabine<- drugsensitivity[-1:-7,]
NegLogGI50gemcitabine2<- NegLogGI50gemcitabine[-2:-8,]

#substract the value 7,003322 for GI50 value of gemcitabine for every cell line 
NegLogGI50gemcitabine3 <- NegLogGI50gemcitabine2 - 7.003322
Neg_LogGI50gemcitabine4 <- as.data.frame(NegLogGI50gemcitabine3)


library(ggplot2)
theme_set(theme_bw())  
Neg_LogGI50gemcitabine4$`cell_line` <- rownames(Neg_LogGI50gemcitabine4)
Neg_LogGI50gemcitabine4$NegLogGI50gemcitabine3_type<- ifelse(Neg_LogGI50gemcitabine4$NegLogGI50gemcitabine3< 0, "below", "above")
ggplot(Neg_LogGI50gemcitabine4, aes(x=`cell_line`, y=NegLogGI50gemcitabine3, label=NegLogGI50gemcitabine3)) + 
    geom_bar(stat='identity', aes(fill=NegLogGI50gemcitabine3_type), width=0.7)  +
    scale_fill_manual(name="value", 
                       labels = c("Above Average", "Below Average"), 
                       values = c("above"="#00ba38", "below"="#f8766d")) + 
     labs(subtitle=" ", 
          title= "cell line and drug sensitivity") + 
coord_flip()

#looking at the previous mean graph plot now compare the differences in mutations of cell lines which had high gemcitabine sensitivity and which had low sensitivity   

#filter out celllines with 10 highest and 10 lowest GI50 values
sort(NegLogGI50gemcitabine2, decreasing = FALSE)

#cell lines with highest GI50 values are SNB-19, OVCAR-8 ,CAKI-1,NCI-H23 ,HCT-116 ,NCI-H460 ,M14,MCF7 ,786-0,ACHN
#since SNB-19 is not part of the mutations data I am going to proceed with the nine other celllines
celllineshigh <- c("OVCAR-8","CAKI-1","NCI-H23","HCT-116","NCI-H460","M14","MCF7","786-0","ACHN")

#cell lines with lowest GI50 values are HCC-2998, HS-578T, OVCAR-4 , TK-10, PC-3, MDA-MB-231, SK-MEL-28 , OVCAR-3 ,KM12, HCT-15
celllineslow <- c("HCC-2998","HS-578T","OVCAR-4","TK-10","PC-3","MDA-MB-231","SK-MEL-28","OVCAR-3","KM12","HCT-15")

#missense mutation for celllines with high GI50 values 
for (i in (celllineshigh) ) {print(sum(mutations[which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] == "Missense_Mutation"))}
missensehigh <- c(292,148,907,2229,317,993,420,285,96)

#missense mutation for celllines with low GI50 values 
for (i in (celllineslow) ) {print(sum(mutations[which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] == "Missense_Mutation"))}
missensenselow <- c(6021,172,334,243,119,357,242,173,2492,6716)

#silent mutation for celllines with high GI50 values 
for (i in (celllineshigh) ) {print(sum(mutations[which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] == "Silent"))}
silenthigh <- c(130,56,387,1039,113,555,153,120,37)

#silent mutation for celllines with low GI50 values 
for (i in (celllineslow) ) {print(sum(mutations[which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] == "Silent"))}
silentlow <- c(1540,49,145,102,45,142,172,82,1305,2635)

#nonsense mutation for celllines with high GI50 values 
for (i in (celllineshigh) ) {print(sum(mutations[which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] == "Nonsense_Mutation"))}
nonsensehigh <- c(17,10,68,113,19,62,30,17,5)

#nonsense mutation for celllines with low GI50 values 
for (i in (celllineslow) ) {print(sum(mutations[which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] == "Nonsense_Mutation"))}
nonesenselow <- c(772,8,5,11,4,22,20,13,135,366)

#splice site mutation for celllines with high GI50 values
for (i in (celllineshigh) ) {print(sum(mutations[which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] == "Splice_Site"))}
splicesitehigh <- c(19,10,66, 161,14,57,20,24,11)

#splice site mutation for celllines with low GI50 values 
for (i in (celllineslow) ) {print(sum(mutations[which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] == "Splice_Site"))}
splicesitelow <- c(187,11,18,12,6,22,26,17,176,541)

#frame shift deletion for celllines with high GI50 values 
for (i in (celllineshigh) ) {print(sum(mutations[which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] == "Frame_Shift_Del"))}
frameshifthigh <- c(16,14,40,631,21,5,9,18,3)

#frame shift deletion for celllines with low GI50 values
for (i in (celllineslow) ) {print(sum(mutations[which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] == "Frame_Shift_Del"))}
frameshiftlow <- c(5,6,6,7,7,18,8,9,459,153)

#frame shift insertion for celllines with high GI50 values 
for (i in (celllineshigh) ) {print(sum(mutations[which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] == "Frame_Shift_Ins"))}
frameshiftinshigh <- c(23,15,22,175,18,2,15,12,14)

#frame shift insertion for celllines with low GI50 values 
for (i in (celllineslow) ) {print(sum(mutations[which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] == "Frame_Shift_Ins"))}
frameshiftinslow <- c(17,14,8,4,8,13,25,12,124,70)

#de novo out of frame mutation for celllines with high GI50 values
for (i in (celllineshigh) ) {print(sum(mutations[which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] == "De_novo_Start_OutOfFrame"))}
denovohigh <- c(3,5,15,20,3,2,0,5,4)

#de novo out of frame mutation for celllines with low GI50 values 
for (i in (celllineslow) ) {print(sum(mutations[which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] == "De_novo_Start_OutOfFrame"))}
denovolow <- c(14,1,3,0,0,6,2,2,32,113)

#create a matrix for high GI50 values 
HighGI50 <- cbind(missensehigh,silenthigh,nonsensehigh,splicesitehigh,frameshifthigh,frameshiftinshigh,denovohigh)
dimnames(HighGI50) = list( 
    c("OVCAR-8","CAKI-1","NCI-H23","HCT-116","NCI-H460","M14","MCF7","786-0","ACHN"),         # row names
    c("missense", "silent", "nonsense","splicesite","frame_shift_deletion","frame_shift_insertion","de_novo_out_of_frame")) #column names
    

#create a matrix for low GI50 values 
LowGI50 <- cbind(missensenselow,silentlow,nonesenselow,splicesitelow,frameshiftlow,frameshiftinslow,denovolow)
dimnames(LowGI50) = list( 
     c("HCC-2998","HS-578T","OVCAR-4","TK-10","PC-3","MDA-MB-231","SK-MEL-28","OVCAR-3","KM12","HCT-15"),         # row names 
     c("missense", "silent", "nonsense","splicesite","frame_shift_deletion","frame_shift_insertion","de_novo_out_of_frame")) #column names

#create heatmaps 
library("pheatmap")

#using quantiles to determine the colour coding breaks for the heatmap

quantile(HighGI50)

pheatmap(HighGI50,color = colorRampPalette(c("blue","white","purple","firebrick3","red","orange"))(10), breaks = c(0,6,19,113,2229))

quantile(LowGI50)

pheatmap(LowGI50,color = colorRampPalette(c("blue","white","purple","firebrick3","red","orange"))(10), breaks = c(0,8,22,172,6716))

``` 

## Linear regression Analysis

The goal of the linear regression is to find a way to predict the drug sensitivity (GI50 value) of celllines, based on their normal gene expression (when untreated) and/or based on their gene copy number. To do this, we will start by making sure that both matrixes (untreated and copy number) have the same rows and columns (same genes and same celllines) to be able to make a reasonable comparison.

Check dimension of untreated_normalized and copynumber to see if they are the same

```{r}
dim(untreated_normalized)
dim(copynumber)
```

Dimensions are completely different:  

1. copynumber lists more genes than untreated_normalized  
Solution:  
  + Check if rows (genes) are ordered alphabetically  
  + If not: order both matrixes alphabetically by row name  
  + Delete rows that appear in only one of the matrixes  

2. untreated_normalized has more columns, since there is one column for each drug  
Solution:  
 + New matrix with the mean for each cellline in untreated_normalized  
 + Order columns alphabetically in both matrixes if not already ordered  

#### Solving problem 1:

COPY NUMBER:

```{r}
identical(rownames(copynumber), rownames(copynumber[order(rownames(copynumber)),]))
```
False, so order them:
```{r}
cnforlr <- copynumber[order(rownames(copynumber)),]
```

UNTREATED: 

```{r}
identical(rownames(untreated_normalized), rownames(untreated_normalized[order(rownames(untreated_normalized)),]))
```
False, so order them:
```{r}
utnforlr <- untreated_normalized[order(rownames(untreated_normalized)),]
```

Keeping only the rows (genes) that are in both matrixes (copy number and untreated):
```{r}
commongenes <- intersect(rownames(cnforlr), rownames(utnforlr))
cnforlr <- cnforlr[which(rownames(cnforlr) %in% commongenes),]
utnforlr <- utnforlr[which(rownames(utnforlr) %in% commongenes),]
```
#### Solving problem 2:

Cell lines in matrix utnforlr are untreated, so the part of the colnames stating drug, time and dose (0 nM) are unnecessary.  
Delete that part of the name to leave only the cellline.  

```{r}
colnames(utnforlr) <- sub("_.*", "", colnames(utnforlr))
```

Order columns alphabetically 
```{r}
utnforlr <- utnforlr[, order(colnames(utnforlr))]  
```
Check dimension of the matrix for untreated cells
```{r}
dim(utnforlr)
colnames(utnforlr[,1:15])
```
There are  many columns for each cell line, so they need to be merged and have their mean calculated
```{r}
utnforlr <- sapply(split(seq_len(ncol(utnforlr)),colnames(utnforlr)),function(cis) rowMeans(utnforlr[,cis,drop=F]))
dim(utnforlr)
colnames(utnforlr[,1:15])
#Number of columns has been reduced and now there is only one for each cellline
```

Order columns alphabetically in the copy number matrix
```{r}
cnforlr <- copynumber[, order(colnames(copynumber))]
```
Keep in both matrixes only the columns (celllines) that are present in both 
```{r}
commoncelllines <- intersect(colnames(cnforlr), colnames(utnforlr))
cnforlr <- cnforlr[, which(colnames(cnforlr) %in% commoncelllines)]
utnforlr <- utnforlr[, which(colnames(utnforlr) %in% commoncelllines)]
```
Now both matrixes have the same rows (genes) and columns (celllines)

### Plotting the regressions

#### Comparing the gene expression of the biomarkers vs. the drug sensitivity

A matrix is created with the gene expression of the biomarkers and a column with the drug sensitivity of each cell line to be used for the scatter plots and to trace the regression line.

```{r}
drugsensitivityt <- t(drugsensitivity)
dsforlr <- drugsensitivityt[which(rownames(drugsensitivityt) %in% commoncelllines),]
utnbmforlr <- utnforlr[which(rownames(utnforlr) %in% biomarkers),]

dsgemcitibine <- dsforlr[, 7]
linearregression <- rbind(utnbmforlr, dsgemcitibine)
linearregression <- t(linearregression)

namesforlr <- rownames(utnbmforlr)
namesforlr <- c(namesforlr, "DS_gemcitibine")
colnames(linearregression) <- namesforlr

DS_gemcitibine <- linearregression[, "DS_gemcitibine"]
```

Plots of each biomarker's gene expression vs the drug sensitivity:

```{r}
for (i in 1:nrow(utnbmforlr)) {
  GeneX <- linearregression[, i]
  name <- namesforlr[i]
  plot(GeneX, DS_gemcitibine, xlab = namesforlr[i], ylab = "Drug sensitivity", main = "Gene expression vs. drug sensitivity")
  abline(lm(DS_gemcitibine~linearregression[,i], data = as.data.frame(linearregression)), col = "red")
}
```

#### Comparing the gene copy number of the biomarkers vs. the drug sensitivity

Preparation of the data for the analysis:

```{r}
cnbmforlr <- cnforlr[which(rownames(cnforlr) %in% biomarkers),]
cnbmforlr <- t(cnbmforlr)

lrforcn <- cbind(cnbmforlr, DS_gemcitibine)
```

Plots of each biomarker's gene copy number vs. the drug sensitivity

```{r}
for (i in 1:ncol(cnbmforlr)) {
  Gene.i.CN <- cnbmforlr[, i]
  name <- namesforlr[i]
  plot(Gene.i.CN, DS_gemcitibine, xlab = namesforlr[i], ylab = "Drug sensitivity", main = "Gene copy number vs. drug sensitivity")
  abline(lm(DS_gemcitibine~lrforcn[,i], data = as.data.frame(lrforcn)), col = "red")
}
```

#### Multiple regression

```{r}
multipleregression <- lm(DS_gemcitibine~ATF3 + GADD45A + CDKN1A + TNFAIP3 + EGR1 + GDF15 + BTG2 + GEM + CEACAM1 + DHRS2 + HBEGF + IL11 + TNFRSF9 + TXNIP, data = as.data.frame(linearregression))
summary(multipleregression)
```

#### Conclusion from the linear regression

It is clear that there is no linear relationship between the gene expression or the gene copy number of the biomarkers and the drug sensitivity of a cellline to gemcitabine, or at least not with the methods used. 
