---
title: "Cellular response to Gemcitabine and other drugs"
author: "Diego Montero Solano, Sophie Zhu, Tabita Häßner, Shivani Nundoo"
date: "6/5/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Inhaltsverzeichnis fehlt noch!!

## Loading the data

Gemcitabine is a nucleoside analog used as chemotherapy. It is marketed as Gemzar. The drug replaces one of the building blocks of nucleic acids, in this case cytidine, during DNA replication. The process arrests tumor growth, as new nucleosides cannot be attached to the "faulty" nucleoside, resulting in apoptosis. Gemcitabine is used in various carcinomas: non-small cell lung cancer, pancreatic cancer, bladder cancer and breast cancer. It is being investigated for use in oesophageal cancer, and is used experimentally in lymphomas and various other tumor types.
```{r}
#Set wd to the working directory where this Rscript is located
wd = getwd()

#Load data
basalexpression <- readRDS(paste0(wd,"/CCLE_basalexpression.rds"))
copynumber <- readRDS(paste0(wd,"/CCLE_copynumber.rds"))
mutations <- readRDS(paste0(wd,"/CCLE_mutations.rds"))
treated <- readRDS(paste0(wd,"/NCI_TPW_gep_treated.rds"))
untreated <- readRDS(paste0(wd,"/NCI_TPW_gep_untreated.rds"))
drugsensitivity <- readRDS(paste0(wd,"/NegLogGI50.rds"))

cannotation <- read.table(paste0(wd,"/cellline_annotation.tsv"), header = TRUE, 
                          sep ="\t", stringsAsFactors = TRUE)
dannotation <- read.table(paste0(wd,"/drug_annotation.tsv"), header = TRUE, 
                          sep ="\t", stringsAsFactors = TRUE)
metadata <- read.table(paste0(wd,"/NCI_TPW_metadata.tsv"), header = TRUE, 
                       sep ="\t", stringsAsFactors = TRUE)
```

## Scaling of data

```{r}
#scale treated & untreated
untreated_normalized <- scale(untreated)
treated_normalized <- scale(treated)
```

## Comparing boxplots

```{r echo=TRUE, results = 'hide', warning=FALSE, message = F}
library("viridis")
```
```{r}

boxplot(treated, main= "Gene Expression treated celllines",
ylab="Gene expression", xlab="Samples", names = FALSE)
```
Numerous batches are observed in the above plot. About 15 of them are counted. This corresponds to the number of drugs used.  
```{r}
palette(viridis(15))
drugnames <- metadata$drug
par(mar=c(5,4,5,9))
boxplot(treated,medcol= "black", border=drugnames, col=drugnames, xlab="samples", 
ylab="gene expression", main="Gene Expression treated celllines", names=FALSE, 
xaxt="n", boxwex=1, boxlty=0)#
levels=as.factor(levels(drugnames))
legend("topright",inset=c(-0.4,0), legend=levels(drugnames),xpd=TRUE,pch=19,
col=levels,title="drugs")

boxplot(treated_normalized, main = "normalized geneexpression after treatment", 
ylab = "geneexpression", xlab = "samples", names = FALSE) 



```


## PCA
PCA is a good method to reduce dimensionality of data. Here we carry out a PCA of the foldchange so the difference after in geneexpression before and after drug treatment. We also want to see whether it is maybe tissue assoicated as well.

```{r}
fcgeneexpression <- treated_normalized - untreated_normalized
pca_FC = prcomp(fcgeneexpression, center = F, scale. = F)

#plot PC 1-4 of FC

drug_as_factor <- as.factor(metadata[1:ncol(treated),"drug"])
levels <- as.factor(levels(drug_as_factor))
palette(rainbow(15))

par(xpd=T, mar=par()$mar+c(0,0,0,6))
plot(pca_FC$rotation[, 1], pca_FC$rotation[, 2], pch = 20, xlab = "PC 1",
ylab = "PC 2", col=drug_as_factor, main = "PC 1&2 of the fold change ")
legend("right",inset = c(-0.3,0), levels(drug_as_factor), xpd = TRUE,
pch=20, col = levels)

par(xpd=T, mar=par()$mar+c(0,0,0,5))
plot(pca_FC$rotation[, 3], pca_FC$rotation[, 4], pch = 20, xlab = "PC 3",
ylab = "PC 4", col=drug_as_factor, main = "PC 3&4 of the fold change ")
legend("right", inset=c(-0.5,0),levels(drug_as_factor), pch=20,
col = levels, xpd =TRUE)

#pca of untreated assigned to tissue

pca_untreated = prcomp(untreated_normalized, center = F, scale. = F)
plot(pca_untreated$rotation[, 1], pca_untreated$rotation[, 2], col = metadata[1:ncol(untreated), "tissue"], pch = 19, xlab = "PC1",
ylab = "PC2")
plot(pca_untreated$rotation[, 1], pca_untreated$rotation[, 2], col = metadata[1:ncol(untreated), "tissue"], pch = 19, xlab = "PC1", 
ylab = "PC2")
legend("right", inset=c(-0.5,0),levels(drug_as_factor), pch=20,
col = levels, xpd =TRUE)

#pca 1&2 of FC tissue associated
pca_FC = prcomp(fcgeneexpression, center = F, scale. = F)
plot(pca_FC$rotation[, 1], pca_FC$rotation[, 2], col = metadata[1:ncol(untreated), "tissue"], pch = 19, xlab = "PC 1", 
ylab = "PC 2")
legend("right", inset=c(-0.5,0),levels(drug_as_factor), pch=20,
col = levels, xpd =TRUE)

```

FC is obviously drug associated. 
## T-test

Normalisation
```{r echo=TRUE, results = 'hide', warning=FALSE, message = F}
library("car")
```
```{r}
grep("gemcitibine",colnames(treated))

Treated_Cells_Gemcitabine= treated[,365:420]
qqPlot(Treated_Cells_Gemcitabine)

grep("gemcitibine",colnames(untreated))

Untreated_Cells_Gemcitabine=untreated[,365:420]
qqPlot(Untreated_Cells_Gemcitabine)
```
Checking if normalization will have any effect on the QQ-Plots
```{r echo=TRUE, results = 'hide', warning=FALSE, message = F}
library(BBmisc)
```
```{r}
Treated_Gemcitabine_Normalised=normalize(Treated_Cells_Gemcitabine)
qqPlot(Treated_Gemcitabine_Normalised)

Untreated_Gemcitabine_Normalised=normalize(Untreated_Cells_Gemcitabine)
qqPlot(Untreated_Gemcitabine_Normalised)
```
A more pronounced linearity is observed
But no significant change in shape, original data is used for further processing.

## T.test

The H0 Hypothesis: there is no change in the mean of the treated and untreated data.
Alternative Hypothesis (H1): there is a difference in mean between the treated and untreated data.
Chosen limit: when pValue < 0.05,  H0 is rejected

```{r}
dim(Untreated_Cells_Gemcitabine)

commondataframe=data.frame(Untreated_Cells_Gemcitabine,Treated_Cells_Gemcitabine)

pValues <- apply(commondataframe, 1, function(x) t.test(x[1:56],x[57:112],paired = TRUE, alternative = "two.sided")$p.value)
commondataframe_pvalue <- data.frame(commondataframe, pValues)
sum(pValues < 0.05)
```
6744 genes had a p value less than 0.05. Thus, this leads us to reject the null hypothesis 
and assume that there could be a difference between the treated and untreated data for these particular genes.


## Biomarkers

In order to asses the effect that gemcitabine has in the cancer cells, we will define certain genes,
that deliver the most information. To this end, we will base the selection of the genes on their
expression, or more specifically, the change in their expression before and after the treatment with the
drug. These will be out biomarkers.  

The first step is to calculate a fold change to find the difference in expression of every single gene between treated and untreated celllines.

```{r}
fcgeneexpression <- treated_normalized - untreated_normalized
#Absolute value of the data in fcgeneexpression, to find the biggest changes in gene expression, regardless of if they are up- or down-regulated
absvalfcgeneexpression <- abs(fcgeneexpression)
```

New matrices containing only the celllines treated with gemcitabine:

```{r, warning = FALSE, message = FALSE}
library("dplyr")

treated_gemcitabine <- select(as_tibble(treated_normalized), contains("gemcitibine"))
treated_gemcitabine <- as.matrix(treated_gemcitabine)
rownames(treated_gemcitabine) <- rownames(treated_normalized)

untreated_gemcitabine <- 
  select(as_tibble(untreated_normalized), contains("gemcitibine"))
untreated_gemcitabine <- as.matrix(untreated_gemcitabine)
rownames(untreated_gemcitabine) <- rownames(untreated_normalized)

absvalfcge_gemcitabine <- 
  select(as_tibble(absvalfcgeneexpression), contains("gemcitibine"))
absvalfcge_gemcitabine <- as.matrix(absvalfcge_gemcitabine)
rownames(absvalfcge_gemcitabine) <- rownames(absvalfcgeneexpression)
```

Create a new matrix with the top differentially expressed genes in descending order

```{r}
orderforvectors <- function(x) {
  y <- names(x[order(-x)])
  return(y)
}

#The new function is applied to the columns of the fold change matrix
matrixofbiomarkers <- apply(absvalfcge_gemcitabine, 2, orderforvectors) 
```

Check for most common genes in the first 15 rows of the new matrix

```{r}
topgenes <- sort(table(as.vector(matrixofbiomarkers[1:15,])))

matrixtopgenes <- as.matrix(topgenes[order(-topgenes)])

barplot(matrixtopgenes[1:30,], xlab = "", ylab = "Counts", 
        main = "Top differentially expressed genes", 
        cex.names = 0.7, las = 2, col = "#00CC66")
mtext("Genes", side = 1, line = 4)
```
We decided to take the top 15 genes with the highest difference in expression before and after treatment as our biomarkers.

List of our biomarkers:

```{r}
biomarkers <- names(matrixtopgenes[1:15,1])
biomarkers
```

### Further insight into our biomarkers:

Now that we know which genes have the highest difference in gene expression, we will check if each of 
them is up- or down-regulated.

```{r}
fcge_gemcitabine <- 
  select(as_tibble(fcgeneexpression), contains("gemcitibine"))
fcge_gemcitabine <- as.matrix(fcge_gemcitabine)
rownames(fcge_gemcitabine) <- rownames(fcgeneexpression)

fcge_gemcitabine <- fcge_gemcitabine[which(rownames(fcge_gemcitabine) %in% biomarkers),]
#fcge_gemcitabine <- fcge_gemcitabine[biomarkers,]

meanfcge_gemcitabine <- apply(fcge_gemcitabine, 1, mean)
meanfcge_gemcitabine <- meanfcge_gemcitabine[order(-meanfcge_gemcitabine)]

barplot(meanfcge_gemcitabine, xlab = "", ylab = "Fold change", ylim = c(-1, 3), 
        main = "Mean change in gene expression of the biomarkers with gemcitabine", 
        cex.names = 0.9, las = 2, col = "#00CC66")
mtext("Biomarkers", side = 1, line = 4)
```

Apparently most of the biomarkers are up-regulated after the treatment with gemcitabine. Only DHRS2 is down-regulated.

### Cellular function of the biomarkers

To further understand the effects of gemcitabine in the cells, we decided to compile the main functions 
of each biomarker. The information about each gene was taken from the gene database of the National 
Center for Biotechnology Information (NCBI) of the USA.


Biomarker | Description
----------|-------------
ATF3      | This gene encodes for activating transcription factors. It is often highly expressed in cancer cells and is involved in the complex process of cellular stress response.
CXCL8     | It codes for Interleukin-8, which induces the migration of immune cells towards sites of infection, stimulates phagocytisis and promotes angiogenesis.
CDKN1A    | This gene codes for a cyclin-dependent kinase inhibitor, which plays a role in the regulation of the cell cycle, including DNA replication and DNA damage repair. Since gemcitabine produces broken DNA strands, it makes sense that this DNA-damage-repair protein would be up-regulated in cancer cells to counteract the drug's effects.
GADD45A   | Member of a group of genes whose transcription increases when the cell is under stress or being treated with DNA-damaging agents, such as gemcitabine.
TNFAIP3   | Encodes for TNF alpha induced protein 3, which is involved in the cytikine-mediated immune and inflammatory responses.
HBEGF     | Encodes for a growth factor involved in wound healing processes and known to be over-expressed in certain cancer celllines.
GEM       | This gene encodes for a GTP-binding protein, which could have a regulatory function in receptor-mediated signal transduction.
TXNIP     | Encodes for a thioredoxin-binding protein, which inhibits thioreduxin and results in the accumulation of reactive oxigen species in the cell and increased oxidative stress. It is suspected to function as a tumor-suppressor gene. 
BTG2      | Encodes for a protein of the BTG/Tob family, which appears to have antiproliferative effects. 
EGR1      | Encodes for a nuclear protein and transcription regulator; it works as a cancer suppressor gene.
GDF15     | Encodes for a TGF-beta ligand, which regulates gene expression. Increased levels of the protein are linked with inflammation and oxidative stress.
TNFRSF9   | Encodes for a protein of the TNF-receptor superfamily, which contributes to the clonal expansion, survival and development of T-cells.
IL11      | Encodes for Interleukin-11, which stimulates the development of immunoglobulin-dependent B-cells.
CEACAM1   | Encodes for a cell-cell adhesion molecule that plays a role in angiogenesis, apoptosis, tumor suppression and metastasis.
DHRS2     | Encodes for a dehydrogenase/reductase of a family of enzymes known to metabolize different compounds such as prostaglandins, retinoids, lipids and xenobiotics.

#### Conclusion from the biomarkers

Most of the biomarkers found are up-regulated after the cell is treated with gemcitabine. When looking 
at the functions that these genes have, it appears that this up-regulation would be beneficial for the 
treatment of cancer, since several of them are tumor-suppressor genes or have immune-activating effects.
The one biomarker that is down-regulated (DHRS2) is an enzyme that metabolzes xenobiotics, so its 
reduced expression is also beneficial, since it would presumably decrease the speed of degradation of 
gemcitabine.

## Microsatellite instability
MSI is defined as a mistake in MMR (Mismatch Repair system) so if we plot the number of mutations against the celllines, 
MSI-H celllines are supposed to have more mutations, to check this, we do the following plot

```{r echo=TRUE, results = 'hide', warning=FALSE, message = F}
library(plyr)
library(qpcR)
library(ggplot2)
```
```{r}
mutation_cellline <- mutations[,"Tumor_Sample_Barcode"]
counting_mutations <- count(mutation_cellline, vars = NULL, wt_var = NULL)
```

```{r}
#creating a new dataframe with the Celllines and whether the Microsatellite Instablility is high or not. TRUE:High  False:Stable/Low

L=cannotation$Microsatellite_instability_status=="MSI-H"
Cell_Line=data.frame("Cell Line name"= cannotation[,"Cell_Line_Name"],"Microsatellite"= L)
Cell_Line <- Cell_Line[order(Cell_Line$Cell.Line.name),]
drugsensitivity=t(drugsensitivity)

#The the Cell_Line dataframe is directly merged with the Gemcitabine Drug Sensitivity Values.
Cell_Line_no_nas <- na.omit(Cell_Line)
table_of_mutations_na <- qpcR:::cbind.na(counting_mutations, Cell_Line_no_nas)
table_of_mutations <- na.omit(table_of_mutations_na)
mutation_plot <- qplot(

    x = table_of_mutations[,3],

    y = table_of_mutations[,2],

    data = table_of_mutations,

    color = Microsatellite, 

    main = "number of mutations per cellline coloured by MSI status", 

    xlab = "Cell line",ylab = "mutations",
)
mutation_plot + ggtitle("number of mutations per cellline coloured by MSI 
status")+ xlab("Cell Line")+ylab("mutations")+theme(axis.text.x = element_text(angle = 90))

```


It is obvious that MSI status affects the number of mutations, due to mutations in the repair system, but does MSI status also effect the effect of gemcitibine? MSI-H cancers are known to have a better prognosis when it comes to immune activating therapies, because of more unknown antigens, being produced. But does it also effect non-immune mechanisms in the cell? A loss of system which repairs wrong parts could also lead to a worse proliferation of the cell.

```{r, warning=FALSE}
# Plotting the GI50 values relating to Gemcitabine treatment .

drugsensitivity=t(drugsensitivity)

gemcitabine_line=data.frame("Gemcitabine drug sensitivity"= drugsensitivity["gemcitibine",])
everything=cbind(Cell_Line,gemcitabine_line)
```

```{r echo=TRUE, results = 'hide', warning=FALSE, message = F}
library(ggplot2)
```
```{r}
basic=ggplot(everything, aes(Cell.Line.name, Gemcitabine.drug.sensitivity, colour = factor(Microsatellite), shape = factor(Microsatellite) )) +
   geom_point()
basic + ggtitle("Drug Sensitivity of Gemcitabine treated cell lines according to High Microsatellite Instability")+ xlab("Cell Line")+ylab("Drug Sensitivity")+theme(axis.text.x = element_text(angle = 90))
```
a relationship between high drug sensitivity(Gemcitabine) and high microsatellite instability cannot be concluded.

# Heatmap of biomarkers due to MSI status

To answer the question wheter MSI status affects the fold change after gemcitibine treatment, we investigate the fold change of our 15 biomarkers for gemcitibine.

```{r}
#check which rows are for gemcitibine
gmci <- grepl('gemcitibine', colnames(fcgeneexpression))

#rows 365:420(gemcitibine), make dataframe with gemci and biomarkers
colnames(fcgeneexpression) = metadata[1:ncol(treated),"cell"]
biomarkers_geneexpression <- fcgeneexpression[c("ATF3", "CXCL8", "GADD45A","CDKN1A","TNFAIP3", "EGR1", "GDF15", "BTG2", "GEM", "CEACAM1", "DHRS2", "HBEGF","IL11", "TNFRSF9", "TXNIP"),365:420]

#MSI related genes, choosen after research, new matrix for untreated and FC
geneexpression_MSI_related <- fcgeneexpression[c("MLH3", "MLH1", "MSH6","PMS2CL///PMS2", "BATF", "BATF3"), 365:420]
geneexpression_MSI_related_untreated <- untreated_normalized[c("MLH3","MLH1", "MSH6", "BATF", "BATF3"), 365:420]

#which cellines MSI-H or L? Note them 
#View(Cell_Line)

#MSI-H in col = 8,12,11,17,19,27,42,45 of geneexpression MSI related which is only gemcitibine specific

#define new matrices to divide matrices in MSI-H and MSI-L
MSI_H_vector <- c(8,12,11,17,19,27,42,45)
biomarkers_geneexpression_MSIH <- biomarkers_geneexpression[,c(8,12,11,17,19,27,42,45)]
biomarkers_geneexpression_MSIL <- biomarkers_geneexpression[,-c(8,12,11,17,19,27,42,45)]

geneexpression_MSI_related_MSIH <- geneexpression_MSI_related[,c(8,12,11,17,19,27,42,45)]
geneexpression_MSI_related_MSIL <- geneexpression_MSI_related[,-c(8,12,11,17,19,27,42,45)]

geneexpression_MSI_related_untreated_MSIH <- geneexpression_MSI_related_untreated[, c(8,12,11,17,19,27,42,45)]
geneexpression_MSI_related_untreated_MSIL <- geneexpression_MSI_related_untreated[, -c(8,12,11,17,19,27,42,45)]


#packages which makes it possible to plot tow heatmaps next to each other
```
```{r echo=TRUE, results = 'hide', warning=FALSE, message = F}
library("ComplexHeatmap")
library("circlize")
library("colorspace")
library("GetoptLong")
```
Does MSI status effect the expression of biomarkers after gemcitibine treatment?
```{r}
#two heatmaps of FC of biomarkers in MSI-Low and MSI-High celllines

ht_biomarkers_geneexpression_MSIH = Heatmap(biomarkers_geneexpression_MSIH,
name = "FC MSI-High", column_title = "MSI-H", column_names_gp = gpar(fontsize = 7)) 
ht_biomarkers_geneexpression_MSIL = Heatmap(biomarkers_geneexpression_MSIL,
name = "FC MSI-Low", column_title = "MSI-L", column_names_gp = gpar(fontsize = 7))
 
ht_biomarkers_geneexpression_MSIH + ht_biomarkers_geneexpression_MSIL


#no obvious difference between MSI-H and MSI-L when it comes to FC after gemci treatment
#sum mean of FC in MSI-L (48 celllines) and MSI-H (8 celllines)

mean(biomarkers_geneexpression_MSIH) ;mean(biomarkers_geneexpression_MSIL)
```
no significant difference between geneexpression in MSI-H and MSI-L biomarkers, positive prognosis of MSI tumors is probably caused by the recognition of the immune system.

## Drug sensitivity and cell line

```{r}
#is the drug sensitivity for gemcitabine dependend on the cell line? 
#plotting mean graph plot 

#calculation of GI50 mean for gemcitabine
rowMeans(drugsensitivity, na.rm = TRUE, dims = 1)
#GI50 mean value is 7,003322 for gemcitabine
#creating a vector for GI50 values of gemcitabine only 
NegLogGI50gemcitabine<- drugsensitivity[-1:-7,]
NegLogGI50gemcitabine2<- NegLogGI50gemcitabine[-2:-8,]

#substract the value 7,003322 for GI50 value of gemcitabine for every cell line 
GI50value <- NegLogGI50gemcitabine2 - 7.003322
Neg_LogGI50gemcitabine4 <- as.data.frame(GI50value)
```

```{r echo=TRUE, results = 'hide', warning=FALSE, message = F}
library(ggplot2)
```
```{r}
theme_set(theme_bw())  
Neg_LogGI50gemcitabine4$`cell_line` <- rownames(Neg_LogGI50gemcitabine4)
Neg_LogGI50gemcitabine4$GI50value_type<- ifelse(Neg_LogGI50gemcitabine4$GI50value< 0,
"below", "above")
ggplot(Neg_LogGI50gemcitabine4, aes(x=`cell_line`, y=GI50value, label=GI50value)) + 
    geom_bar(stat='identity', aes(fill=GI50value_type), width=0.7)  +
    scale_fill_manual(name="value", 
                       labels = c("Above Average", "Below Average"), 
                       values = c("above"="#00ba38", "below"="#f8766d")) + 
     labs(subtitle=" ", 
          title= "Cell line and drug sensitivity") + 
coord_flip()

#looking at the previous mean graph plot now compare the differences in mutations of cell lines which had high gemcitabine sensitivity and which had low sensitivity   

#filter out celllines with 10 highest and 10 lowest GI50 values
sort(NegLogGI50gemcitabine2, decreasing = FALSE)

#cell lines with highest GI50 values are SNB-19, OVCAR-8 ,CAKI-1,NCI-H23 ,HCT-116,
#NCI-H460 ,M14,MCF7 ,786-0,ACHN
#since SNB-19 is not part of the mutations data I am going to proceed with the
#nine other celllines
celllineshigh <- c("OVCAR-8","CAKI-1","NCI-H23","HCT-116","NCI-H460","M14",
"MCF7","786-0","ACHN")

#cell lines with lowest GI50 values are HCC-2998, HS-578T, OVCAR-4 , TK-10, PC-3, 
#MDA-MB-231, SK-MEL-28 , OVCAR-3 ,KM12, HCT-15
celllineslow <- c("HCC-2998","HS-578T","OVCAR-4","TK-10","PC-3","MDA-MB-231",
"SK-MEL-28","OVCAR-3","KM12","HCT-15")

#missense mutation for celllines with high GI50 values 
for (i in (celllineshigh) ) {print(sum(mutations[
which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] == 
"Missense_Mutation"))}
missensehigh <- c(292,148,907,2229,317,993,420,285,96)

#missense mutation for celllines with low GI50 values 
for (i in (celllineslow) ) {print(sum(mutations[
which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] 
== "Missense_Mutation"))}
missensenselow <- c(6021,172,334,243,119,357,242,173,2492,6716)

#silent mutation for celllines with high GI50 values 
for (i in (celllineshigh) ) {print(sum(mutations[
which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"]
== "Silent"))}
silenthigh <- c(130,56,387,1039,113,555,153,120,37)

#silent mutation for celllines with low GI50 values 
for (i in (celllineslow) ) {print(sum(mutations[
which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] 
== "Silent"))}
silentlow <- c(1540,49,145,102,45,142,172,82,1305,2635)

#nonsense mutation for celllines with high GI50 values 
for (i in (celllineshigh) ) {print(sum(mutations[
which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] 
== "Nonsense_Mutation"))}
nonsensehigh <- c(17,10,68,113,19,62,30,17,5)

#nonsense mutation for celllines with low GI50 values 
for (i in (celllineslow) ) {print(sum(mutations[
which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"]
== "Nonsense_Mutation"))}
nonesenselow <- c(772,8,5,11,4,22,20,13,135,366)

#splice site mutation for celllines with high GI50 values
for (i in (celllineshigh) ) {print(sum(mutations[
which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] 
== "Splice_Site"))}
splicesitehigh <- c(19,10,66, 161,14,57,20,24,11)

#splice site mutation for celllines with low GI50 values 
for (i in (celllineslow) ) {print(sum(mutations[
which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"]
== "Splice_Site"))}
splicesitelow <- c(187,11,18,12,6,22,26,17,176,541)

#frame shift deletion for celllines with high GI50 values 
for (i in (celllineshigh) ) {print(sum(mutations[
which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"]
== "Frame_Shift_Del"))}
frameshifthigh <- c(16,14,40,631,21,5,9,18,3)

#frame shift deletion for celllines with low GI50 values
for (i in (celllineslow) ) {print(sum(mutations[
which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"]
== "Frame_Shift_Del"))}
frameshiftlow <- c(5,6,6,7,7,18,8,9,459,153)

#frame shift insertion for celllines with high GI50 values 
for (i in (celllineshigh) ) {print(sum(mutations[
which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] 
== "Frame_Shift_Ins"))}
frameshiftinshigh <- c(23,15,22,175,18,2,15,12,14)

#frame shift insertion for celllines with low GI50 values 
for (i in (celllineslow) ) {print(sum(mutations[
which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"]
== "Frame_Shift_Ins"))}
frameshiftinslow <- c(17,14,8,4,8,13,25,12,124,70)

#de novo out of frame mutation for celllines with high GI50 values
for (i in (celllineshigh) ) {print(sum(mutations[
which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"]
== "De_novo_Start_OutOfFrame"))}
denovohigh <- c(3,5,15,20,3,2,0,5,4)

#de novo out of frame mutation for celllines with low GI50 values 
for (i in (celllineslow) ) {print(sum(mutations[
which(mutations$Tumor_Sample_Barcode == i),"Variant_Classification"] 
== "De_novo_Start_OutOfFrame"))}
denovolow <- c(14,1,3,0,0,6,2,2,32,113)

#create a matrix for high GI50 values 
HighGI50 <- cbind(missensehigh,silenthigh,nonsensehigh,splicesitehigh,
frameshifthigh,frameshiftinshigh,denovohigh)
dimnames(HighGI50) = list( 
    c("OVCAR-8","CAKI-1","NCI-H23","HCT-116","NCI-H460","M14","MCF7","786-0","
    ACHN"),       
    c("missense", "silent", "nonsense","splicesite","frame_shift_deletion",
    "frame_shift_insertion","de_novo_out_of_frame"))    

#create a matrix for low GI50 values 
LowGI50 <- cbind(missensenselow,silentlow,nonesenselow,splicesitelow,
frameshiftlow,frameshiftinslow,denovolow)
dimnames(LowGI50) = list( 
     c("HCC-2998","HS-578T","OVCAR-4","TK-10","PC-3","MDA-MB-231","SK-MEL-28",
     "OVCAR-3","KM12","HCT-15"),        
     c("missense", "silent", "nonsense","splicesite","frame_shift_deletion",
     "frame_shift_insertion","de_novo_out_of_frame")) 
```
create heatmaps 
```{r echo=TRUE, results = 'hide', warning=FALSE, message = F}
library("pheatmap")
```
```{r}
#using quantiles to determine the colour coding breaks for the heatmap

quantile(HighGI50, c(.10, .20, .30, .40, .50, .60, .70, .80, .90, 1))

pheatmap(HighGI50,color = colorRampPalette(c("blue","white","purple",
"firebrick3","red","orange"))(10), breaks = c(4.8,10.7,15.0,19.0,26.4,66.6,149.0,
390.3,2229.0), main= "Celllines with high GI50 fold change values")



quantile(LowGI50, c(.10, .20, .30, .40, .50, .60, .70, .80, .90, 1))

pheatmap(LowGI50,color = colorRampPalette(c("blue","white","purple","firebrick3",
"red","orange"))(10), breaks = c(4.0,6.8,10.4,14.0,22.0,74.8,142.9,
198.0,564.1,6716.0), main= "Celllines with low GI50 fold change values")

```


GI50 is the concentration for 50% of maximal inhibition of cell proliferation.
However, since our values are -log10 transformed, higher values indicate higher 
sensitivity.The most prominent difference in mutation pattern between the celllines 
with high and low gemcitabine sensitivity is that the celllines with a low GI50 value
have a much higher  count of mutations. Further we are going to look at the portion
of different mutation types. In the heatmap it is visible that for most mutation 
type the mutation count pattern is similar. However, for nonsense and splice site
mutation you can see that the celllines with low GI50 values have a higher portion 
in those mutations in comparison to the celllines with high GI50 values. It is
maybe linked to the fact that gemcitabine is an antimetabolite and stops DNA synthesis 
by inducing a masked chain termination. Nonsense mutations have the same effect of 
inducing a stop of DNA synthesis. Celllines in which a stop of DNA synthesis occur 
more often due to nonsense mutation are maybe therefore less sensitive to gemcitabine. 
However mutation frequency alone is insufficient to identify and predict the drug 
sensitivity of a cellline towards gemcitabine. 

Other possibilities to explain a celllines sensitivity to gemcitabine are 
mutations of genes that play an important role in the mechanisms gemcitabine 
is involved in. One enzyme that plays an important role in converting gemcitabine 
from the prodrug into the final active drug form is deoxycytidine kinase (DCK) . 
It modifies gemcitabine by attaching a phosphate to it, converting it to gemcitabine
monophosphate. The gene encoding this enzyme is called DCK.


```{r}

print(mutations[which(mutations$Hugo_Symbol == "DCK"),"Tumor_Sample_Barcode"])

```


The only cellline that has a DCK mutation is "HCC-2998" and it 
corresponds to the fact that it is also a cellline with a low gemcitabine sensitivity. 



## Linear regression Analysis

As seen previously, the analyzed celllines have different sensitivities to the treatment with
gemcitabine. It would be useful to be able to predict the cell's response to the drug before starting
the treatment and be able to apply this knowledge in the real world. The ultimate goal would be to be
able to predict the response of a patient's cancer to the drug, so as to better decide if it is the
ideal therapy in each case.

This linear regression will attempt to find a relation between:  
1. The drug sensitivity of the celllines (GI50 value) and the gene expression of their biomarkers when
untreated.  
2. The drug sensitivity and the gene copy number of the biomarkers.  

The first step is to check the correlation between the gene expression and gene copy number of the
biomarkers, to decide if it makes sense to compare them separately to the drug sensitivity or if it is
enough to compare only one of the two.

##### Creating a drug sensitivity and a gene expression matrix containing only the biomarkers:

The biomarkers were selected from the treated and untreated matrices, so first we make sure that they
are in the copy number matrix too.

```{r}
biomarkers %in% rownames(copynumber)
```
The second biomarker is not in this matrix, so it has to be removed from the gene expression one as well.

```{r}
commonbiomarkers.lr <- intersect(rownames(copynumber), biomarkers)
#There is no need to intersect with the untreated matrix, because it was used to define 
#the biomarkers, so it definitely contains all of them.
```

Keep only the common biomarkers in both matrices and order alphabetically:

```{r}
cnforlr <- copynumber[order(rownames(copynumber)),]
cnforlr <- cnforlr[which(rownames(cnforlr) %in% commonbiomarkers.lr),]

untreatedforlr <- untreated_normalized[order(rownames(untreated_normalized)),]
untreatedforlr <- 
  untreatedforlr[which(rownames(untreatedforlr) %in% commonbiomarkers.lr),]
```

Check the dimensions of the matrices:

```{r}
dim(cnforlr)
dim(untreatedforlr)

colnames(untreatedforlr[,1:3])
```

The matrix for untreated has many more columns because it contains a measurement for each cellline
before starting the treatment with each drug. Since no drug has been applied to any of them, all columns
regarding a given cellline can be merged, and the information of drug, dose and time in the column name
can be eliminated.

```{r}
colnames(untreatedforlr) <- sub("_.*", "", colnames(untreatedforlr))

untreatedforlr <- sapply(split(seq_len(ncol(untreatedforlr)),colnames(untreatedforlr)),
                         function(cis) rowMeans(untreatedforlr[,cis,drop=F]))
dim(untreatedforlr)
```

The matrix for untreated still lists more columns than the gene copy number, so only the common ones are
to be kept, and we order them alphabetically:

```{r}
commoncelllines.lr <- intersect(colnames(untreatedforlr), colnames(cnforlr))

untreatedforlr <- untreatedforlr[,which(colnames(untreatedforlr) %in% commoncelllines.lr)]
untreatedforlr <- untreatedforlr[,order(colnames(untreatedforlr))]

cnforlr <- cnforlr[,which(colnames(cnforlr) %in% commoncelllines.lr)]
cnforlr <- cnforlr[,order(colnames(cnforlr))]
```

##### Check the correlation between the gene expression of untreated cell lines and their gene copy number

The Spearman correlation of the copy number and gene expression of each biomarker is calculated.

```{r}
cnforlr.t <- t(cnforlr)
untreatedforlr.t <- t(untreatedforlr)

correlations.cn.genexp <- c()

for (i in 1:ncol(cnforlr.t)) {
  cor.i <- cor(cnforlr.t[,i], untreatedforlr.t[,i], method = "spearman")
  correlations.cn.genexp <- c(correlations.cn.genexp, cor.i)
}
names(correlations.cn.genexp) <- colnames(untreatedforlr.t)

barplot(correlations.cn.genexp, ylim = c(-1,1), las = 2, cex.names = 0.8, col = "#3399FF", 
        ylab = "Correlation coefficient", 
        main = "Correlation between gene expression and copy number" )
mtext("Genes", side = 1, line = 4.3)
```

The correlation coefficient of all biomarkers' gene copy number and gene expression are very close to 0,
which means that there is a very low correlation between the two variables analyzed.  

This low correlation was expected,since there are several mechanisms that control cells' gene
expression. Epigenetic variations in the different tissues, like the DNA methylation can silence the
gene expression, and transcription factors also play an important role in the up- or down-regulation of
gene expression. Therefore, taking only the gene copy number would harldy be enough to predict the
expression of a given gene.  

Since the correlation of the two variables is so low, we will proceed with separate linear regressions for:  
1. Gene expression vs. drug sensitivity  
2. Gene copy number vs. drug sensitivity

### 1. Linear regression of gene expression vs. drug sensitivity

The drug sensitivity matrix contains the EC50 values of several celllines when exposed to different
drugs. First of all, we make sure to leave only the celllines that are contained in the other two
matrices for the regression.

```{r}
dsforlr <- drugsensitivity[,which(colnames(drugsensitivity) %in% commoncelllines.lr)]
dsforlr <- dsforlr[,order(colnames(dsforlr))]
dsforlr <- dsforlr[order(rownames(dsforlr)),]
```

Create a matrix with the gene expression of the biomarkers and a row with the drug sensitivity of each
cellline, to be used for the linear regressions:

```{r}
genexp.drugsens <- rbind(untreatedforlr, dsforlr["gemcitibine",])
rownames(genexp.drugsens) <- c(rownames(genexp.drugsens)[1:nrow(genexp.drugsens)-1], 
                               "DS_Gemcitabine")
genexp.drugsens <- t(genexp.drugsens)
```

Linear regression of the first biomarker (ATF3):

```{r}
lrATF3genexp <- lm(DS_Gemcitabine~ATF3, data = as.data.frame(genexp.drugsens))
summary(lrATF3genexp)
plot(DS_Gemcitabine~ATF3, data = as.data.frame(genexp.drugsens), 
     ylab = "EC50", main = "Linear regression with gene expression and EC50")
abline(lm(DS_Gemcitabine~ATF3, data = as.data.frame(genexp.drugsens)), col = "#3399FF")
```

R-squared values of the linear regression of each biomarker:

```{r}
DS_gemcitabine <- genexp.drugsens[,"DS_Gemcitabine"]
all.r.squared.ge <- c()

for (i in 1:length(commonbiomarkers.lr)) {
  r.squared.ge.i <- 
    summary(lm(DS_gemcitabine~genexp.drugsens[,i], 
               data = as.data.frame(genexp.drugsens)))$r.squared
  all.r.squared.ge <- c(all.r.squared.ge, r.squared.ge.i)
}

names(all.r.squared.ge) <- colnames(genexp.drugsens[,1:ncol(genexp.drugsens)-1])

barplot(all.r.squared.ge, ylim = c(0,1), ylab = "R-squared value", 
        main = "R-squared values of the linear regressions using gene expression", 
        cex.names = 0.7, las = 2, col = "#3399FF")
mtext("Biomarkers", side = 1, line = 4)
```

### 2. Linear regression of gene copy number vs. drug sensitivity

The first step is to create a matrix with the gene copy number of the biomarkers and a row with the drug
sensitivity of each cellline, just like it was done with the gene expression:

```{r}
copynum.drugsens <- rbind(cnforlr, dsforlr["gemcitibine",])
rownames(copynum.drugsens) <- c(rownames(copynum.drugsens)[1:nrow(copynum.drugsens)-1], 
                               "DS_Gemcitabine")
copynum.drugsens <- t(copynum.drugsens)
```

Linear regression of the first biomarker (ATF3):

```{r}
lrATF3copynum <- lm(DS_Gemcitabine~ATF3, data = as.data.frame(copynum.drugsens))
summary(lrATF3copynum)
plot(DS_Gemcitabine~ATF3, data = as.data.frame(copynum.drugsens),
     ylab = "EC50", main = "Linear regression with gene copy number and EC50")
abline(lm(DS_Gemcitabine~ATF3, data = as.data.frame(copynum.drugsens)), col = "#3399FF")
```

R-squared values of the linear regression of each biomarker:

```{r}
all.r.squared.cn <- c()
for (i in 1:length(commonbiomarkers.lr)) {
  r.squared.i.cn <- 
    summary(lm(DS_gemcitabine~copynum.drugsens[,i], 
               data = as.data.frame(copynum.drugsens)))$r.squared
  all.r.squared.cn <- c(all.r.squared.cn, r.squared.i.cn)
}
names(all.r.squared.cn) <- colnames(copynum.drugsens[,1:ncol(copynum.drugsens)-1])
barplot(all.r.squared.cn, ylim = c(0,1), ylab = "R-squared value", 
        main = "R-squared values of the linear regressions using gene copy number", 
        cex.names = 0.7, las = 2, col = "#3399FF")
mtext("Biomarkers", side = 1, line = 4)
```

## Multiple regression analysis

The linear regressions showed that the gene expression or copy number of each biomarker by itself does
not account for the drug sensitivity of the cellline. Therefore, we will inspect if a multiple
regression using all 15 biomarkers can be used to better predict a cellline's drug sensitivity.

#### Gene expression vs. drug sensitivity

```{r}
multipleregression.ge <- lm(DS_Gemcitabine~ATF3 + GADD45A + CDKN1A + TNFAIP3 + EGR1 
                         + GDF15 + BTG2 + GEM + CEACAM1 + DHRS2 + HBEGF + IL11 + 
                           TNFRSF9 + TXNIP, data = as.data.frame(genexp.drugsens))
summary(multipleregression.ge)
```

#### Gene copy number vs. drug sensitivity

```{r}
multipleregression.cn <- lm(DS_Gemcitabine~ATF3 + GADD45A + CDKN1A + TNFAIP3 + EGR1 
                         + GDF15 + BTG2 + GEM + CEACAM1 + DHRS2 + HBEGF + IL11 + 
                           TNFRSF9 + TXNIP, data = as.data.frame(copynum.drugsens))
summary(multipleregression.cn)
```

This regression has the highest R-squared value of all, but at 0.53 it still cannot be used to predict
the drug sensitivity of the celllines.
