---
title: "Cellular response to Gemcitabine and other drugs"
author: "Diego Montero Solano, Sophie Zhu, Tabita Häßner, Shivani Nundoo"
date: "6/5/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##Loading the data

```{r}
#Set wd to the working directory where this Rscript is located
wd = getwd()

#Load data
basalexpression <- readRDS(paste0(wd,"/CCLE_basalexpression.rds"))
copynumber <- readRDS(paste0(wd,"/CCLE_copynumber.rds"))
mutations <- readRDS(paste0(wd,"/CCLE_mutations.rds"))
treated <- readRDS(paste0(wd,"/NCI_TPW_gep_treated.rds"))
untreated <- readRDS(paste0(wd,"/NCI_TPW_gep_untreated.rds"))
drugsensitivity <- readRDS(paste0(wd,"/NegLogGI50.rds"))

cannotation <- read.table(paste0(wd,"/cellline_annotation.tsv"), header = TRUE, sep ="\t", stringsAsFactors = TRUE)
dannotation <- read.table(paste0(wd,"/drug_annotation.tsv"), header = TRUE, sep ="\t", stringsAsFactors = TRUE)
metadata <- read.table(paste0(wd,"/NCI_TPW_metadata.tsv"), header = TRUE, sep ="\t", stringsAsFactors = TRUE)
```

## Normalization of data

```{r}
#loading library limma needed for normalization
library(limma)

#normalize treated & untreated
untreated_normalized <- normalizeBetweenArrays(untreated)
treated_normalized <- normalizeBetweenArrays(treated)
```

##Comparing boxplots

```{r}

#finding number of batches 
boxplot(treated)
boxplot(treated_normalized) 
```
Though there seems to be large numbers of samples grouped together forming about 11 clusters. We cannot conclude that there is a batch effect as the x-axis does not have any sample names( may be due to the number of samples, 819). WE do not know exactly which samples are present in each group. The clustering can simply be due to the effect of the drug or natural causes. IMPORTANT: Moreover, further operations were carried out without the consideration or removal of the batch effects (if any).

##PCA

```{r}
#PCA of FC 
fcgeneexpression <- treated_normalized - untreated_normalized
pca_FC = prcomp(fcgeneexpression, center = F, scale. = F)

#plotting PCA of FC drug associated, PCA 1-6

plot(pca_FC$rotation[, 1], pca_FC$rotation[, 2], col = metadata[1:ncol(untreated), "drug"], pch = 19, xlab = "PC 1", 
ylab = "PC 2")

plot(pca_FC$rotation[, 3], pca_FC$rotation[, 4], col = metadata[1:ncol(untreated), "drug"], pch = 19, xlab = "PC 3", 
ylab = "PC 4")

plot(pca_FC$rotation[, 5], pca_FC$rotation[, 6], col = metadata[1:ncol(untreated), "drug"], pch = 19, xlab = "PC 5", 
ylab = "PC 6")
```

#FC is obviously drug associated

##T-test

Check if normalized treated data is normally distributed using QQ-plot
If normally distributed quantile dots should lie on a straight line 

```{r}
d = sample(1:nrow(treated_normalized), 1)
qqnorm(treated_normalized[d, ], main = rownames(treated_normalized)[d])
qqline(treated_normalized[d, ])
#check if normaized untreated data is normally distributed using QQ-plot
c = sample(1:nrow(untreated_normalized), 1)
qqnorm(untreated_normalized[c, ], main = rownames(untreated_normalized)[c])
qqline(untreated_normalized[c, ])
```

T-test with normalized data? Significant difference between treated and untreated?

```{r}
t.test(treated, untreated, paired = TRUE)
t.test(treated_normalized, untreated_normalized, paired = TRUE)
```

The H0 Hypothesis is that there is no change in the mean of the treated and untreated data.

The 95 percent confidence interval is between -0.0015656142 and -0.0009451127 with a p-value of 2.182e-15.
The p-value is not within the 95 percent confidence interval, thus the null hypothesis is rejected for the alternative hypothesis.

Alternative Hypothesis: there is a difference in mean between the treated and untreated data. ("Tentatively": does this prove that the change is provoked by the drug? or is it only a batch effect? Check the boxplots for batch effect identification.

##Biomarkers

Fold change to find the difference in gene expression between treated and untreated celllines 

```{r}
fcgeneexpression <- treated_normalized - untreated_normalized

#Absolute value of the data in fcgeneexpression, to find the biggest changes, regardless of if they are up- or down-regulated
absvalfcgeneexpression <- abs(fcgeneexpression)
```

New matrices with only the celllines treated wiht gemcitabine

```{r}
# ***CAMBIAR Y HACER CON SPLIT FUNCTION***
treated_gemcitabine <- treated_normalized[,365:420]
untreated_gemcitabine <- untreated_normalized[,365:420]
absvalfcge_gemcitabine <- absvalfcgeneexpression[,365:420]
```

Create new function to get a list of the genes with the biggest difference in expression after treatment for a single column (cell line) of absvalfcge_gembitabine:

```{r}
orderforbiomarkers <- function(absvalfcge_gemcitabine, n) {
  gfc <- absvalfcge_gemcitabine[ , n]
  orderedgenes <- names(gfc[order(-gfc)])
  return(orderedgenes)
}
```

Function to order the components of a vector (descending) and return the names in that order

```{r}
orderforvectors <- function(x) {
  y <- names(x[order(-x)])
  return(y)
}
```

Apply the new funtion *orderforvectors* to all the columns and create a new matrix with the gene names

```{r}
matrixofbiomarkers <- apply(absvalfcge_gemcitabine, 2, orderforvectors) 
#the 2 means it is applied to columns
```

Check for most common genes in the first 15 rows of the new matrix

```{r}
topgenes <- sort(table(as.vector(matrixofbiomarkers[1:15,])))
```

Convert *topgenes* into a matrix in descending order of counts

```{r}
matrixtopgenes <- as.matrix(topgenes[order(-topgenes)])
```

Bar plot of *matrixtopgenes* to find the biomarkers

```{r}
barplot(matrixtopgenes[1:20,], xlab = "", ylab = "Counts", main = "Top differentially expressed genes", cex.names = 0.7, las = 2)
mtext("Genes", side = 1, line = 4)
#We decided to take the top 15 genes as our biomarkers
```

List of our biomarkers

```{r}
biomarkers <- names(matrixtopgenes[1:15,1])
biomarkers
```
