---
title: "Cellular response to Gemcitabine and other drugs"
author: "Diego Montero Solano, Sophie Zhu, Tabita Häßner, Shivani Nundoo"
date: "6/5/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Loading the data

```{r}
#Set wd to the working directory where this Rscript is located
wd = getwd()

#Load data
basalexpression <- readRDS(paste0(wd,"/CCLE_basalexpression.rds"))
copynumber <- readRDS(paste0(wd,"/CCLE_copynumber.rds"))
mutations <- readRDS(paste0(wd,"/CCLE_mutations.rds"))
treated <- readRDS(paste0(wd,"/NCI_TPW_gep_treated.rds"))
untreated <- readRDS(paste0(wd,"/NCI_TPW_gep_untreated.rds"))
drugsensitivity <- readRDS(paste0(wd,"/NegLogGI50.rds"))

cannotation <- read.table(paste0(wd,"/cellline_annotation.tsv"), header = TRUE, sep ="\t", stringsAsFactors = TRUE)
dannotation <- read.table(paste0(wd,"/drug_annotation.tsv"), header = TRUE, sep ="\t", stringsAsFactors = TRUE)
metadata <- read.table(paste0(wd,"/NCI_TPW_metadata.tsv"), header = TRUE, sep ="\t", stringsAsFactors = TRUE)
```

## Normalization of data

```{r}
#loading library limma needed for normalization
library(limma)

#normalize treated & untreated
untreated_normalized <- normalizeBetweenArrays(untreated)
treated_normalized <- normalizeBetweenArrays(treated)
```

## Comparing boxplots

```{r}
boxplot(treated, main = "geneexpression after treatment", ylab = "geneexpression", xlab = "samples", names = FALSE)
boxplot(treated_normalized, main = "normalized geneexpression after treatment", ylab = "geneexpression", xlab = "samples", names = FALSE) 
```


## PCA

```{r}
#PCA of FC 
fcgeneexpression <- treated_normalized - untreated_normalized
pca_FC = prcomp(fcgeneexpression, center = F, scale. = F)

#plot PC 1-4 of FC

drug_as_factor <- as.factor(metadata[1:ncol(treated),"drug"])
levels <- as.factor(levels(drug_as_factor))

par(xpd=T, mar=par()$mar+c(0,0,0,6))
plot(pca_FC$rotation[, 1], pca_FC$rotation[, 2], pch = 20, xlab = "PC 1",ylab = "PC 2", col=drug_as_factor, main = "PC 1&2 of the fold change ")
legend("right",inset = c(-0.3,0), levels(drug_as_factor), xpd = TRUE, pch=20, col = levels)

par(xpd=T, mar=par()$mar+c(0,0,0,5))
plot(pca_FC$rotation[, 3], pca_FC$rotation[, 4], pch = 20, xlab = "PC 3",ylab = "PC 4", col=drug_as_factor, main = "PC 3&4 of the fold change ")
legend("right", inset=c(-0.5,0),levels(drug_as_factor), pch=20, col = levels, xpd =TRUE)
```

FC is obviously drug associated

## T-test

Check if normalized treated data is normally distributed using QQ-plot
If normally distributed quantile dots should lie on a straight line 

```{r}
#check if normalized treated data is normally distributed using QQ-plot
qqnorm(treated_normalized, main = "QQ Plot of geneexpression after treatment")
qqline(treated_normalized)
```

T-test with normalized data? Significant difference between treated and untreated?

```{r}
t.test(treated, untreated, paired = TRUE)
t.test(treated_normalized, untreated_normalized, paired = TRUE)
```

The H0 Hypothesis is that there is no change in the mean of the treated and untreated data.

The 95 percent confidence interval is between -0.0015656142 and -0.0009451127 with a p-value of 2.182e-15.
The p-value is not within the 95 percent confidence interval, thus the null hypothesis is rejected for the alternative hypothesis.

Alternative Hypothesis: there is a difference in mean between the treated and untreated data. ("Tentatively": does this prove that the change is provoked by the drug? or is it only a batch effect? Check the boxplots for batch effect identification.

## Biomarkers

Fold change to find the difference in gene expression between treated and untreated celllines 

```{r}
fcgeneexpression <- treated_normalized - untreated_normalized

#Absolute value of the data in fcgeneexpression, to find the biggest changes, regardless of if they are up- or down-regulated
absvalfcgeneexpression <- abs(fcgeneexpression)
```

New matrices with only the celllines treated wiht gemcitabine

```{r}
library("dplyr")

treated_gemcitabine <- select(as_tibble(treated_normalized), contains("gemcitibine"))
treated_gemcitabine <- as.matrix(treated_gemcitabine)
rownames(treated_gemcitabine) <- rownames(treated_normalized)

untreated_gemcitabine <- select(as_tibble(untreated_normalized), contains("gemcitibine"))
untreated_gemcitabine <- as.matrix(untreated_gemcitabine)
rownames(untreated_gemcitabine) <- rownames(untreated_normalized)

absvalfcge_gemcitabine <- select(as_tibble(absvalfcgeneexpression), contains("gemcitibine"))
absvalfcge_gemcitabine <- as.matrix(absvalfcge_gemcitabine)
rownames(absvalfcge_gemcitabine) <- rownames(absvalfcgeneexpression)
```


Create a function to order the components of a vector (descending) and return the names in that order

```{r}
orderforvectors <- function(x) {
  y <- names(x[order(-x)])
  return(y)
}
```

Apply the new funtion *orderforvectors* to all the columns and create a new matrix with the gene names

```{r}
matrixofbiomarkers <- apply(absvalfcge_gemcitabine, 2, orderforvectors) 
#the 2 means it is applied to columns
```

Check for most common genes in the first 15 rows of the new matrix

```{r}
topgenes <- sort(table(as.vector(matrixofbiomarkers[1:15,])))
```

Convert *topgenes* into a matrix in descending order of counts

```{r}
matrixtopgenes <- as.matrix(topgenes[order(-topgenes)])
```

Bar plot of *matrixtopgenes* to find the biomarkers

```{r}
barplot(matrixtopgenes[1:20,], xlab = "", ylab = "Counts", main = "Top differentially expressed genes", cex.names = 0.7, las = 2)
mtext("Genes", side = 1, line = 4)
```
We decided to take the top 15 genes as our biomarkers

List of our biomarkers:

```{r}
biomarkers <- names(matrixtopgenes[1:15,1])
biomarkers
```
## Microsatellite instability

```{r}
# Plotting the GI50 values relating to Gemcitabine treatment .

plot(drugsensitivity["gemcitibine",])

#creating a new dataframe with the Celllines and whether the Microsatellite Instablility is high or not. TRUE:High  False:Stable/Low

L=cannotation$Microsatellite_instability_status=="MSI-H"
Cell_Line=data.frame("Cell Line name"= cannotation[1:61,1],"Microsatellite"= L)
Cell_Line <- Cell_Line[order(Cell_Line$Cell.Line.name),]
drugsensitivity=t(drugsensitivity)

# It is observed that the cell lines in the drugsensitivity matrix is already arranged alphabetically.
# Thus the Cell_Line dataframe is directly merged with the Gemcitabine Drug Sensitivity Values.
```
gemcitabine_line=data.frame("Gemcitabine drug sensitivity"= drugsensitivity[1:61,8])
everything=cbind(Cell_Line,gemcitabine_line)
plot(everything) #Quite interesting to look at
```{r}
gemcitabine_line=data.frame("Gemcitabine drug sensitivity"= drugsensitivity[1:61,8])
everything=cbind(Cell_Line,gemcitabine_line)

library(ggplot2)
basic=ggplot(everything, aes(Cell.Line.name, Gemcitabine.drug.sensitivity, colour = factor(Microsatellite), shape = factor(Microsatellite) )) +
   geom_point()
basic #very satisfied with this graph

# a relationship between high drug sensitivity(Gemcitabine) and high microsatellite instability cannot be concluded.
```

#Heatmap of biomarkers and of MSI related genes 
```{r}
#check which rows are for gemcitibine
gmci <- grepl('gemcitibine', colnames(fcgeneexpression))

#rows 365:420, make dataframe with gemci and biomarkers
biomarkers_geneexpression <- fcgeneexpression[c("ATF3", "CXCL8", "GADD45A", "CDKN1A", "TNFAIP3", "EGR1", "GDF15", "BTG2", "GEM", "CEACAM1", "DHRS2", "HBEGF", "IL11", "TNFRSF9", "TXNIP"),365:420]

#MSI related genes, choosen after research, new matrix for untreated and FC
geneexpression_MSI_related <- fcgeneexpression[c("MLH3", "MLH1", "MSH6", "PMS2CL///PMS2", "BATF", "BATF3"), 365:420]
geneexpression_MSI_related_untreated <- untreated_normalized[c("MLH3", "MLH1", "MSH6", "PMS2CL///PMS2", "BATF", "BATF3"), 365:420]

#make colnames equal to celllines
colnames(biomarkers_geneexpression) = metadata[365:420, 2]
colnames(geneexpression_MSI_related) <- metadata[365:420,2]
colnames(geneexpression_MSI_related_untreated) <- metadata[365:420,2]

#which cellines MSI-H or L? Note them 
#View(Cell_Line)

#MSI-H in col = 8,12,11,17,19,27,42,45 of geneexpression MSI related which is only gemcitibine specific


#define new matrices to divide matrices in MSI-H and MSI-L

biomarkers_geneexpression_MSIH <- biomarkers_geneexpression[,c(8,12,11,17,19,27,42,45)]
biomarkers_geneexpression_MSIL <- biomarkers_geneexpression[,-c(8,12,11,17,19,27,42,45)]

geneexpression_MSI_related_MSIH <- geneexpression_MSI_related[,c(8,12,11,17,19,27,42,45)]
geneexpression_MSI_related_MSIL <- geneexpression_MSI_related[,-c(8,12,11,17,19,27,42,45)]

geneexpression_MSI_related_untreated_MSIH <- geneexpression_MSI_related_untreated[, c(8,12,11,17,19,27,42,45)]
geneexpression_MSI_related_untreated_MSIL <- geneexpression_MSI_related_untreated[, -c(8,12,11,17,19,27,42,45)]


#packages which makes it possible to plot tow heatmaps next to each other
```
```{r echo=TRUE, results = 'hide', warning=FALSE, message = F}
library("ComplexHeatmap")
library("circlize")
library("colorspace")
library("GetoptLong")
```
Does MSI status effect the expression of biomarkers after gemcitibine treatment?
```{r}
#two heatmaps of FC of biomarkers in MSI-Low and MSI-High celllines
ht_biomarkers_geneexpression_MSIH = Heatmap(biomarkers_geneexpression_MSIH, name = "FC MSI-High", column_title = "MSI-H", column_names_gp = gpar(fontsize = 7)) 
ht_biomarkers_geneexpression_MSIL = Heatmap(biomarkers_geneexpression_MSIL, name = "FC MSI-Low", column_title = "MSI-L", column_names_gp = gpar(fontsize = 7))
 
ht_biomarkers_geneexpression_MSIH + ht_biomarkers_geneexpression_MSIL


#no obvious difference between MSI-H and MSI-L when it comes to FC after gemci treatment
#sum mean of FC in MSI-L (48 celllines) and MSI-H (8 celllines)

sum(biomarkers_geneexpression_MSIH)/8 ; sum(biomarkers_geneexpression_MSIL)/48
t.test(biomarkers_geneexpression_MSIH, biomarkers_geneexpression_MSIL, paired = TRUE)

#no significant difference between geneexpression in MSI-H and MSI-L biomarkers

#######Plotting mutations######
#compare mean of mutations, biomarkers_geneexpression_MSIH and L has to be changed, MSIH contains only MSIH celllines
mean_mutation_MSIH <- sum(mutations[,16] %in% colnames(biomarkers_geneexpression_MSIH))/8
mean_mutation_MSIL <- sum(mutations[,16] %in% colnames(biomarkers_geneexpression_MSIL))/48
```
